Update the Ethereum bridge specs (#36)
anoma/namada-docs: specs/src/interoperability/ethereum-bridge/ethereum_events_attestation.md

# Ethereum Events Attestation

We want to store events from the smart contracts of our bridge onto
Namada. We will include events that have been voted by at least one
validator, but will not act on them until they have been `seen` by a
weighted average of more than $\frac{2}{3}$ of the voting power
available across all epochs of the tally. Generically, we consider some

> tally $t$ `seen` if it satisfies the following inequation:

Reviewer: ```suggestion
tally $t$ `seen` if it satisfies the following inequality:
```

}
```

Validators may never vote more than once on a given event. To ensure
that this invariant is held, we keep track of who voted on some event
and events are timed out if they are not `seen` within the span of
`unbonding_length` epochs, which corresponds to the period of time
necessary for bonded tokens to be returned to an address (check the
[relevant proof-of-stake
section](../../economics/proof-of-stake/reward-distribution.md) for more
details). Timing out an event consists of removing all its associated
state from storage. Therefore, this mechanism serves another purpose:
purging forged events from storage, voted on by Byzantine validators.

## Minimum confirmations

There will be a protocol-specified minimum number of confirmations that
events must reach on the Ethereum chain, before validators can vote to
include them on Namada. This minimum number of confirmations will be
changeable via governance.

`TransferToNamada` events may include a custom minimum number of
confirmations that must be at least the protocol-specified minimum
number of confirmations. However, this value is initially set to
**100**.

> Validators must not vote to include events that have not met the

Reviewer: ```suggestion
Validators must NOT vote to include events that have *yet* to meet the
```

required number of confirmations. Votes on unconfirmed events will
eventually time out in storage, unless the number of confirmations was

> only off by a few block heights in Ethereum. Assuming that an honest

Author: is that regarding "unless the number of confirmations was only off by a few block heights in Ethereum"?

> only off by a few block heights in Ethereum. Assuming that an honest

Reviewer: Yes correct, how does this work in practice? How many heights are "a few"?

```

Where `$msg_hash` is the SHA256 digest of the Borsh serialization of
some `EthereumEvent`.

Changes to this `/eth_msgs` storage subspace are only ever made by nodes
as part of the ledger code based on the aggregate of votes by validators
for specific events. That is, changes to `/eth_msgs` happen in block `n`
in a deterministic manner based on the votes included in the block
proposal for block `n`. Depending on the underlying Tendermint version,
these votes will either be included as vote extensions or as protocol
transactions.

The `/eth_msgs` storage subspace will belong to the `EthBridge` validity
predicate. It should disallow any changes to this storage from wasm
transactions.

### Including events into storage

For every Namada block proposal, the proposer should include the votes
for events from other validators into their proposal. If the underlying
Tendermint version supports vote extensions, consensus invariants
guarantee that a quorum of votes from the previous block height can be
included. Otherwise, validators can only submit votes by broadcasting
protocol transactions, which comes with less guarantees (i.e. no

> consensus finality).

Reviewer 2: My understanding of protocol transactions is that they are final, but that a proposer can choose to include or not include votes and protocol transactions independently, while with ABCI++ a proposer must include either the vote and the vote extension or neither. Is that accurate, or not? If it's accurate, we should clarify a bit.

Author: so, to clarify, a vote is just an abstract concept - in these cases, a vote can be made through a vote extension, or through a protocol transaction. a vote extension piggy backs off of BFT consensus, while a protocol transaction must be recorded on-chain, and go through the mempool (hence the delay in the latter).

you could claim that a proposer can censor votes, by not accounting for them, but this is also the case with votes in the sense of BFT consensus (each protocol step can be considered a vote that advances the state of consensus).

regarding protocol transactions being final in the sense of BFT consensus, I don't think they are, unless you do some weird on-chain 3-step consensus.

Reviewer 2: hmm, I think I might be using the terms slightly differently. Protocol transactions, if included by some proposer in a block, are broadcast to the other validators, who execute them as part of that block, updating the state root accordingly, yes? That's all I mean by "finality" - once a protocol transaction is included in a block accepted by consensus, the state changes are final just as for regular transactions.

the difference which I understood - which might be wrong - is that I thought Tendermint had validators sign over the vote extension data in the vote itself, such that the proposer would have to choose one of two options:
1. don't include the (BFT consensus) vote, and don't include the vote extension data
2. include the BFT consensus vote, and include the vote extension data

with protocol transactions, the proposer has two more options:
3. include the BFT consensus vote, but _don't_ include the "vote extension" (protocol transaction) data
4. don't include the BFT consensus vote, but _do_ include the "vote extension" (protocol transaction) data

am I misunderstanding something here?

> The vote of a validator should include the events of the Ethereum blocks
> they have seen via their full node such that:

Reviewer 2: ```suggestion
The vote of a validator should include each event of the Ethereum blocks their full node has seen that satisfies the following conditions:
```

}
```

These votes will be delivered to subsequent block proposers who will
aggregate those that they can verify and will inject them into their
proposal. With ABCI++ this involves creating a new protocol transaction,

> we dub a digest, comprised of multiple individual votes on Ethereum

Reviewer 2: ```suggestion
dubbed a _digest_, comprised of multiple individual votes on Ethereum
```
